utils/wasm-gen/src/utils.rs

```diff
@@ -215,6 +222,172 @@ fn find_recursion_impl<Callback>(
     path.pop();
 }
 
+pub fn inject_critical_gas_limit(module: Module, critical_gas_limit: u64) -> Module {
```

Comment:
the function is really huge. Can you slice it and add more docs?

Reply:
done in 00bf3602a70ad63d0f7a1afa14297c52946765a9